ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/FrameBasedAnimationDriver.java

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.react.uimanager.animation;
+
+import com.facebook.react.bridge.ReadableArray;
+import com.facebook.react.bridge.ReadableMap;
+
+/**
+ * Implementation of {@link AnimationDriver} which provides a support for simple time-based
+ * animations that are pre-calculate on the JS side. For each animation frame JS provides a value
+ * from 0 to 1 that indicates a progress of the animation at that frame.
+ */
+class FrameBasedAnimationDriver extends AnimationDriver {
+
+  private long mStartFrameTimeNanos = -1;
+  private final double[] mFrames;
+  private final double mToValue;
+  private double mFromValue;
+  private boolean mHasToValue;
```

Comment:
Why do you have this variable? You always set toValue and toValue is required.

One suggestion: it would greatly simplify the implementation if you always added the toValue as last element of the array. This an easy way to solve the problem of the last element not ending at the exact end value.